In [77]:
import pandas as pd
import numpy as np

In [78]:
df_base = pd.read_csv("data/baseline_values.csv", index_col=False, header=None,
                      names=['name', 'baseline_value'])
# set metric names to lower case
df_base['name']=df_base['name'].map(lambda x: x.lower())


In [79]:
# Rename for easier read
df_base['metric'] = ['cookies', 'clicks', 'user-ids', 'ctp', 'gross conversion', 'retention', 'net conversion']
df_base['dmin'] = [3000, 240, -50, 0.01, -0.01, 0.01, 0.0075]
df_base

name  baseline_value  \
0  unique cookies to view course overview page pe...    40000.000000   
1  unique cookies to click "start free trial" per...     3200.000000   
2                               enrollments per day:      660.000000   
3   click-through-probability on "start free trial":        0.080000   
4             probability of enrolling, given click:        0.206250   
5              probability of payment, given enroll:        0.530000   
6                probability of payment, given click        0.109313   

             metric       dmin  
0           cookies  3000.0000  
1            clicks   240.0000  
2          user-ids   -50.0000  
3               ctp     0.0100  
4  gross conversion    -0.0100  
5         retention     0.0100  
6    net conversion     0.0075

### 3. Measure of variability in metrics
3.1 Scaling

In [80]:
n_sample = 5000
n_cookie = df_base.loc[df_base['metric']=='cookies', 'baseline_value'].iloc[0]
scaling_factor = n_sample/n_cookie

# create a new column to store scaled baseline values
df_base['scaled_n']= np.nan
df_base['scaled_n'] = (df_base['baseline_value'] * scaling_factor).where(df_base['metric'].isin(['cookies', 'clicks', 'user-ids']), np.nan) 
df_base 

name  baseline_value  \
0  unique cookies to view course overview page pe...    40000.000000   
1  unique cookies to click "start free trial" per...     3200.000000   
2                               enrollments per day:      660.000000   
3   click-through-probability on "start free trial":        0.080000   
4             probability of enrolling, given click:        0.206250   
5              probability of payment, given enroll:        0.530000   
6                probability of payment, given click        0.109313   

             metric       dmin  scaled_n  
0           cookies  3000.0000    5000.0  
1            clicks   240.0000     400.0  
2          user-ids   -50.0000      82.5  
3               ctp     0.0100       NaN  
4  gross conversion    -0.0100       NaN  
5         retention     0.0100       NaN  
6    net conversion     0.0075       NaN

3.2 Standard error

$SE=\sqrt{\frac{\hat p(1-\hat p)}{n}}$

In [81]:
def stderr(m_n, m_p):
    #calculate standard error of the population
    #Start with df_base. m_n = metric of unit of diversion, m_p = metric of probability 
    f1 = df_base[df_base['metric'] == m_n]
    if f1.empty:
        raise ValueError(f'Metric {m_n} not found.')

    f2 = df_base[df_base['metric'] == m_p]
    if f2.empty:
        raise ValueError(f'Metric {m_p} not found.')
    
    n = f1['scaled_n'].iloc[0]
    p = f2['baseline_value'].iloc[0]
    return (p * (1 - p) / n) ** 0.5


# Gross Conversion (clicks)
se_gross_conv = stderr('clicks', 'gross conversion')

# Retention (enrolls)
se_retention = stderr('user-ids', 'retention')

# Net conversion (clicks)
se_net_conv = stderr('clicks', 'net conversion')

print(f'Gross Conversion (clicks): {se_gross_conv: .4f}, Retenstion (enrolls): {se_retention:.4f}, Net conversion (clicks): {se_net_conv:.4f}')

Gross Conversion (clicks):  0.0202, Retenstion (enrolls): 0.0549, Net conversion (clicks): 0.0156


In [82]:
# create a new column to store standard error for evaluation metrics
df_base['se'] = [np.nan, np.nan, np.nan, np.nan, round(se_gross_conv, 4), round(se_retention, 4), round(se_net_conv, 4)]
df_base

name  baseline_value  \
0  unique cookies to view course overview page pe...    40000.000000   
1  unique cookies to click "start free trial" per...     3200.000000   
2                               enrollments per day:      660.000000   
3   click-through-probability on "start free trial":        0.080000   
4             probability of enrolling, given click:        0.206250   
5              probability of payment, given enroll:        0.530000   
6                probability of payment, given click        0.109313   

             metric       dmin  scaled_n      se  
0           cookies  3000.0000    5000.0     NaN  
1            clicks   240.0000     400.0     NaN  
2          user-ids   -50.0000      82.5     NaN  
3               ctp     0.0100       NaN     NaN  
4  gross conversion    -0.0100       NaN  0.0202  
5         retention     0.0100       NaN  0.0549  
6    net conversion     0.0075       NaN  0.0156

3.3 Determine sample size

Calculate the sample size for the selected evaluation metrics, use alpha = 0.05 and beta = 0.2.

In [93]:
from statsmodels.stats.power import zt_ind_solve_power
import math 
alpha=0.05
beta =0.2

In [84]:
def spl_size(m):
    # Calculate sample size.
    # Start with df_base. m = metric for prob and mde 
    f3 = df_base[df_base['metric'] == m]
    if f3.empty:
        raise ValueError(f'Metric {m} not found.')
    p = f3['baseline_value'].iloc[0]
    mde = f3['dmin'].iloc[0]

    # pooled probability of p1 and p2
    pooled_prob = (p+(p+abs(mde)))/2
    # calculate effective size using minimum detectable effect and pooled proability
    effect_size = abs(mde)/((pooled_prob*(1-pooled_prob))**0.5)
    return 2*zt_ind_solve_power(effect_size=effect_size, alpha=alpha, power=1-beta, alternative='two-sided')

In [ ]:
# clicks for clicks for gross conversion
n_sample_gross_conv = spl_size('gross conversion')
# Enrollment for clicks for retention
n_sample_retention = spl_size('retention')
# clicks for clicks for net conversion
n_sample_net_conv = spl_size('net conversion')

print(f'Clicks needed for Gross Conversion: {n_sample_gross_conv: .0f}, Retenstion: {n_sample_retention:.0f}, Net conversion: {n_sample_net_conv:.0f}')


Clicks needed for Gross Conversion:  52312, Retenstion: 78104, Net conversion: 55970


Use the sample size calculator (https://www.evanmiller.org/ab-testing/sample-size.html)

Eg Gross Coversion:

Given Baseline Value =  20.625%, and Minimum Detectable Effect = 0.01, using the sample size calculator, the sample size for one group for the number of clicks on "start free trial" is 25,835. Since our A/B testing experiment have one control group and one experiment group, the sample size for the clicks n = 25835*2=51,670.

- Clicks for Gross Conversion = 25835*2=51670
- Enrollment for Retention = 39115*2=78230
- Clicks for Net Conversion = 27413*2=54826

Set the pageviews based on number of clicks. Pageviews = 40000, clicks = 3200, so the ratio of clicks/pageviews = 3200/40000=0.08.
Set the pageviews based on number of enrollments. Pageviews = 40000, enrollments = 660, so the ratio of enrollments/pageviews = 660/40000=0.0165.

- Pageviews for Gross Conversion = 51670/0.08 = 645875
- Pageviews for Retention = 78230/0.0165 = 4741212
- Pageviews for Net Conversion = 54826/0.08 = 685325

In order to test for all three metrics, use the maximum number of pageviews, 4741212.

3.4 Experiment duration and exposure

What fraction of Udacity's traffic would you divert to this experiment?

If we divert 100% of the traffic, given 40,000 pageviews per day, the experiment will take 119 days. This is too long for an experiment. If we eliminate retention and only care about gross conversion and net conversion, and divert 100% of the traffic, we only need 18 days. If we divert 80% of traffic to the experiment, we need 22 days.

In [94]:
n_pageviews = 4741212
duration = n_pageviews/n_cookie
print(math.ceil(duration))

119


In [95]:
n_pageviews = 685325
duration = n_pageviews/n_cookie
print(math.ceil(duration))

18


In [96]:
n_pageviews = 685325
duration = n_pageviews/(n_cookie*0.8)
print(math.ceil(duration))

22


### 4. Sanity Check
For each metric that you chose as an invariant metric, compute a 95% confidence interval for the value you expect to observe. Enter the upper and lower bounds, and the observed value, all to 4 decimal places.

In [97]:
df_cont = pd.read_csv('data/control.csv')
df_cont

Date  Pageviews  Clicks  Enrollments  Payments
0   Sat, Oct 11       7723     687        134.0      70.0
1   Sun, Oct 12       9102     779        147.0      70.0
2   Mon, Oct 13      10511     909        167.0      95.0
3   Tue, Oct 14       9871     836        156.0     105.0
4   Wed, Oct 15      10014     837        163.0      64.0
5   Thu, Oct 16       9670     823        138.0      82.0
6   Fri, Oct 17       9008     748        146.0      76.0
7   Sat, Oct 18       7434     632        110.0      70.0
8   Sun, Oct 19       8459     691        131.0      60.0
9   Mon, Oct 20      10667     861        165.0      97.0
10  Tue, Oct 21      10660     867        196.0     105.0
11  Wed, Oct 22       9947     838        162.0      92.0
12  Thu, Oct 23       8324     665        127.0      56.0
13  Fri, Oct 24       9434     673        220.0     122.0
14  Sat, Oct 25       8687     691        176.0     128.0
15  Sun, Oct 26       8896     708        161.0     104.0
16  Mon, Oct 27       9535     759        233.0     124.0
17  Tue, Oct 28       9363     736        154.0      91.0
18  Wed, Oct 29       9327     739        196.0      86.0
19  Thu, Oct 30       9345     734        167.0      75.0
20  Fri, Oct 31       8890     706        174.0     101.0
21   Sat, Nov 1       8460     681        156.0      93.0
22   Sun, Nov 2       8836     693        206.0      67.0
23   Mon, Nov 3       9437     788          NaN       NaN
24   Tue, Nov 4       9420     781          NaN       NaN
25   Wed, Nov 5       9570     805          NaN       NaN
26   Thu, Nov 6       9921     830          NaN       NaN
27   Fri, Nov 7       9424     781          NaN       NaN
28   Sat, Nov 8       9010     756          NaN       NaN
29   Sun, Nov 9       9656     825          NaN       NaN
30  Mon, Nov 10      10419     874          NaN       NaN
31  Tue, Nov 11       9880     830          NaN       NaN
32  Wed, Nov 12      10134     801          NaN       NaN
33  Thu, Nov 13       9717     814          NaN       NaN
34  Fri, Nov 14       9192     735          NaN       NaN
35  Sat, Nov 15       8630     743          NaN       NaN
36  Sun, Nov 16       8970     722          NaN       NaN

In [98]:
df_exp = pd.read_csv('data/experiment.csv')
df_exp

Date  Pageviews  Clicks  Enrollments  Payments
0   Sat, Oct 11       7716     686        105.0      34.0
1   Sun, Oct 12       9288     785        116.0      91.0
2   Mon, Oct 13      10480     884        145.0      79.0
3   Tue, Oct 14       9867     827        138.0      92.0
4   Wed, Oct 15       9793     832        140.0      94.0
5   Thu, Oct 16       9500     788        129.0      61.0
6   Fri, Oct 17       9088     780        127.0      44.0
7   Sat, Oct 18       7664     652         94.0      62.0
8   Sun, Oct 19       8434     697        120.0      77.0
9   Mon, Oct 20      10496     860        153.0      98.0
10  Tue, Oct 21      10551     864        143.0      71.0
11  Wed, Oct 22       9737     801        128.0      70.0
12  Thu, Oct 23       8176     642        122.0      68.0
13  Fri, Oct 24       9402     697        194.0      94.0
14  Sat, Oct 25       8669     669        127.0      81.0
15  Sun, Oct 26       8881     693        153.0     101.0
16  Mon, Oct 27       9655     771        213.0     119.0
17  Tue, Oct 28       9396     736        162.0     120.0
18  Wed, Oct 29       9262     727        201.0      96.0
19  Thu, Oct 30       9308     728        207.0      67.0
20  Fri, Oct 31       8715     722        182.0     123.0
21   Sat, Nov 1       8448     695        142.0     100.0
22   Sun, Nov 2       8836     724        182.0     103.0
23   Mon, Nov 3       9359     789          NaN       NaN
24   Tue, Nov 4       9427     743          NaN       NaN
25   Wed, Nov 5       9633     808          NaN       NaN
26   Thu, Nov 6       9842     831          NaN       NaN
27   Fri, Nov 7       9272     767          NaN       NaN
28   Sat, Nov 8       8969     760          NaN       NaN
29   Sun, Nov 9       9697     850          NaN       NaN
30  Mon, Nov 10      10445     851          NaN       NaN
31  Tue, Nov 11       9931     831          NaN       NaN
32  Wed, Nov 12      10042     802          NaN       NaN
33  Thu, Nov 13       9721     829          NaN       NaN
34  Fri, Nov 14       9304     770          NaN       NaN
35  Sat, Nov 15       8668     724          NaN       NaN
36  Sun, Nov 16       8988     710          NaN       NaN

In [99]:
# compute the total of cookies, clicks, enrollments, and payments
sums = {'control': pd.Series([df_cont['Pageviews'].sum(), df_cont['Clicks'].sum(), df_cont['Enrollments'].sum(), df_cont['Payments'].sum()],
                             index = ['cookies','clicks','enrollments','payments']),
        'experiment': pd.Series([df_exp['Pageviews'].sum(), df_exp['Clicks'].sum(), df_exp['Enrollments'].sum(), df_exp['Payments'].sum()],
                             index = ['cookies','clicks','enrollments','payments']),
        }

df_sums = pd.DataFrame(sums)
df_sums

control  experiment
cookies      345543.0    344660.0
clicks        28378.0     28325.0
enrollments    3785.0      3423.0
payments       2033.0      1945.0

Count metrics (invariant metrics): expect the total number of cookies and clicks in the control group and the experiment group to be 50% each. SE = sqrt(p(1-p)/n)

In [103]:
df_sums['total']=df_sums['control']+df_sums['experiment']
df_sums['prob']=0.5
# stderr = sqrt(p(1-p)/(n_cont+n_exp))
df_sums['stderr']=np.sqrt((df_sums['prob']*(1-df_sums['prob']))/df_sums['total'])
df_sums['CI_lower']=df_sums['prob']-1.96*df_sums['stderr']
df_sums['CI_upper']=df_sums['prob']+1.96*df_sums['stderr']
df_sums['observed_p']=round(df_sums['experiment']/df_sums['total'],4)
df_sums['Pass Sanity']=df_sums.apply(lambda x: (x.observed_p > x.CI_lower) and (x.observed_p<x.CI_upper), axis=1)
df_sums['d'] = abs((df_sums['control']-df_sums['experiment'])/df_sums['total'])
df_sums

control  experiment     total  prob    stderr  CI_lower  \
cookies      345543.0    344660.0  690203.0   0.5  0.000602  0.498820   
clicks        28378.0     28325.0   56703.0   0.5  0.002100  0.495884   
enrollments    3785.0      3423.0    7208.0   0.5  0.005889  0.488457   
payments       2033.0      1945.0    3978.0   0.5  0.007928  0.484462   

             CI_upper  observed_p  Pass Sanity         d  
cookies      0.501180      0.4994         True  0.001279  
clicks       0.504116      0.4995         True  0.000935  
enrollments  0.511543      0.4749        False  0.050222  
payments     0.515538      0.4889         True  0.022122

To perform sanity check for the Click Through Probability, we would expect that the difference between the two groups be zero.
![standard error for two proportions](images/stderr_calc.png)

In [125]:
# CTP = # of clicks/# of pageviews
df_ctp = pd.DataFrame(
    {
        'ctp_cont': [df_sums['control'].loc['clicks']/df_sums['control'].loc['cookies']],
        'ctp_exp': [df_sums['experiment'].loc['clicks']/df_sums['experiment'].loc['cookies']],
        'n_cont':[df_sums['control'].loc['clicks']]
    }, index=['ctp']
)
df_ctp['se']=np.sqrt(df_ctp['ctp_cont']*(1-df_ctp['ctp_cont'])/df_ctp['n_cont'])
df_ctp['CI_lower'] = df_ctp['ctp_cont'] - 1.96*df_ctp['se']
df_ctp['CI_upper'] = df_ctp['ctp_cont'] + 1.96*df_ctp['se']
df_ctp['Pass Sanity'] = df_ctp.apply(lambda x: (x['ctp_exp']> x['CI_lower']) and (x['ctp_exp']< x['CI_upper']), axis=1)
df_ctp

ctp_cont   ctp_exp   n_cont       se  CI_lower  CI_upper  Pass Sanity
ctp  0.082126  0.082182  28378.0  0.00163  0.078931   0.08532         True

In [123]:
# CTP = # of clicks/# of pageviews
df_ctp = pd.DataFrame(
    {
        'ctp_cont': [df_sums['control'].loc['clicks']/df_sums['control'].loc['cookies']],
        'ctp_exp': [df_sums['experiment'].loc['clicks']/df_sums['experiment'].loc['cookies']],
        'n_cont':[df_sums['control'].loc['clicks']],
        'n_exp':[df_sums['experiment'].loc['clicks']]
    }, index=['ctp']
)
df_ctp['se']=np.sqrt(df_ctp['ctp_cont']*(1-df_ctp['ctp_cont'])/df_ctp['n_cont'] + df_ctp['ctp_exp']*(1-df_ctp['ctp_exp'])/df_ctp['n_exp'])
df_ctp['CI_lower'] = 0 - 1.96*df_ctp['se']
df_ctp['CI_upper'] = 0 + 1.96*df_ctp['se']
df_ctp['observed_p'] = df_ctp['ctp_exp'] - df_ctp['ctp_cont']
df_ctp['Pass Sanity'] = df_ctp.apply(lambda x: (x['observed_p']> x['CI_lower']) and (x['observed_p']< x['CI_upper']), axis=1)
df_ctp

ctp_cont   ctp_exp   n_cont    n_exp        se  CI_lower  CI_upper  \
ctp  0.082126  0.082182  28378.0  28325.0  0.002306  -0.00452   0.00452   

     observed_p  Pass Sanity  
ctp    0.000057         True

### 5. Effect Size Tests
Evaluation metrics: As we use gross conversion and net conversion as the evaluation metrics, we know that d_min=0.01 for gross conversion, and d_min=0.075 for net conversion.

In [ ]:
# Compute the observed difference of GC and NC after excluding nan values
df_cont_e = df_cont.dropna(subset=['Enrollments'])
df_exp_e = df_exp.dropna(subset=['Enrollments'])
sums_e = {'control': pd.Series([df_cont_e['Pageviews'].sum(), df_cont_e['Clicks'].sum(), df_cont_e['Enrollments'].sum(), df_cont_e['Payments'].sum()],
                             index = ['cookies','clicks','enrollments','payments']),
        'experiment': pd.Series([df_exp_e['Pageviews'].sum(), df_exp_e['Clicks'].sum(), df_exp_e['Enrollments'].sum(), df_exp_e['Payments'].sum()],
                             index = ['cookies','clicks','enrollments','payments']),
        }

df_sums_e = pd.DataFrame(sums_e)
df_sums_e['total']=df_sums_e['control']+df_sums_e['experiment']
df_sums_e


control  experiment     total
cookies      212163.0    211362.0  423525.0
clicks        17293.0     17260.0   34553.0
enrollments    3785.0      3423.0    7208.0
payments       2033.0      1945.0    3978.0

In [135]:
# control values
click_cont = df_sums_e.loc['clicks'].control
enroll_cont = df_sums_e.loc['enrollments'].control
payment_cont = df_sums_e.loc['payments'].control

# experiment values
click_exp = df_sums_e.loc['clicks'].experiment
enroll_exp = df_sums_e.loc['enrollments'].experiment
payment_exp = df_sums_e.loc['payments'].experiment

# evaluation metrics
gross_conv_cont = enroll_cont/click_cont
gross_conv_exp = enroll_exp/click_exp

net_conv_cont = payment_cont/click_cont
net_conv_exp = payment_exp/click_exp

gross_conv_pool = (enroll_cont+enroll_exp)/(click_cont+click_exp)
net_conv_pool = (payment_cont+payment_exp)/(click_cont+click_exp)

print('Pooled Gross Conversion: {} \n Pooled Net Conversion:{}'.format(gross_conv_pool,net_conv_pool))


Pooled Gross Conversion: 0.20860706740369866 
 Pooled Net Conversion:0.1151274853124186


In [152]:
def stats(p, n_cont, n_exp, zscore, diff):
    # output standard error, margin, and confidence interval
    # p: pooled probabiliry
    # n_cont, n_exp: number of units in control and experiment
    # diff: evaluation metric different between experiment and control
    stderr = np.sqrt(p*(1-p)*(1/n_cont+1/n_exp))
    ci_lower = diff - zscore*stderr
    ci_upper = diff + zscore*stderr
    return stderr, ci_lower, ci_upper

In [186]:
dmin_gc = -0.01
dmin_nc = 0.0075

zscore = 1.96

In [172]:
d_obs_gc = gross_conv_exp - gross_conv_cont
se_gross, cil_gross, ciu_gross = stats(gross_conv_pool, click_cont, click_exp, zscore, d_obs_gc)

print(click_cont, click_exp, gross_conv_pool, d_obs_gc, se_gross, cil_gross, ciu_gross)

17293.0 17260.0 0.20860706740369866 -0.020554874580361565 0.004371675385225936 -0.0291233583354044 -0.01198639082531873


In [173]:
d_obs_nc = net_conv_exp - net_conv_cont
se_net, cil_net, ciu_net = stats(net_conv_pool, click_cont, click_exp, zscore, d_obs_nc)

print(click_cont, click_exp, net_conv_pool, d_obs_nc, se_net, cil_net, ciu_net)

17293.0 17260.0 0.1151274853124186 -0.0048737226745441675 0.0034341335129324238 -0.011604624359891718 0.001857179010803383


In [187]:
# compute a dataframe to contain the evaluation metrics
effect_size = pd.DataFrame(columns=['dmin_practical','observed diff','CI_lower', 'CI_upper', 'Practically Significant','Statistically Significant'], 
                           index=['Gross Conversion', 'Net Conversion'])

effect_size['dmin_practical'] = [dmin_gc, dmin_nc]
effect_size['observed diff'] = [d_obs_gc, d_obs_nc]
effect_size['CI_lower'] = [cil_gross, cil_net]
effect_size['CI_upper'] = [ciu_gross, ciu_net]
effect_size

dmin_practical  observed diff  CI_lower  CI_upper  \
Gross Conversion         -0.0100      -0.020555 -0.029123 -0.011986   
Net Conversion            0.0075      -0.004874 -0.011605  0.001857   

                 Practically Significant Statistically Significant  
Gross Conversion                     NaN                       NaN  
Net Conversion                       NaN                       NaN

- An evaluation metric is statistically significant if the CI does not include 0 (that is, you can be confident there was a change)
- An evaluation metric is practically significant if the CI does not include the practical significance boundary (that is, you can be confident there is a change that matters to the business.)

In [190]:
effect_size['Practically Significant'] = np.where((effect_size['dmin_practical']>effect_size['CI_lower']) & 
                                                  (effect_size['dmin_practical']<effect_size['CI_upper']), 
                                                  'No', 'Yes')
effect_size['Statistically Significant']= np.where((effect_size['CI_lower']<0 ) &
                                                   (effect_size['CI_upper']>0),
                                                   'No', 'Yes')
effect_size

dmin_practical  observed diff  CI_lower  CI_upper  \
Gross Conversion         -0.0100      -0.020555 -0.029123 -0.011986   
Net Conversion            0.0075      -0.004874 -0.011605  0.001857   

                 Practically Significant Statistically Significant  
Gross Conversion                     Yes                       Yes  
Net Conversion                       Yes                        No

Gross Conversion is both Statistically and Practically significant while Net Conversion is only Practically Significant.


### 6. Sign Test
Run a sign test on each of your evaluation metrics using the day-by-data data. Enter each p-value, and indicate whether each result is statistically significant.

In [ ]:
# If a unit has higher gross conversion in Experiment than in Control , it is a correct sign.
df_sign = pd.merge(df_cont_e[['Date']], df_exp_e[['Date']], on='Date') 
df_sign['GC_cont'] = df_cont_e['Enrollments']/df_cont_e['Clicks']
df_sign['GC_exp'] = df_exp_e['Enrollments']/df_exp_e['Clicks']
df_sign['GC_diff'] = df_sign['GC_exp'] - df_sign['GC_cont']

df_sign['NC_cont'] = df_cont_e['Payments']/df_cont_e['Clicks']
df_sign['NC_exp'] = df_exp_e['Payments']/df_exp_e['Clicks']
df_sign['NC_diff'] = df_sign['NC_exp'] - df_sign['NC_cont']

df_sign

Date   GC_cont    GC_exp   GC_diff   NC_cont    NC_exp   NC_diff
0   Sat, Oct 11  0.195051  0.153061 -0.041990  0.101892  0.049563 -0.052330
1   Sun, Oct 12  0.188703  0.147771 -0.040933  0.089859  0.115924  0.026065
2   Mon, Oct 13  0.183718  0.164027 -0.019691  0.104510  0.089367 -0.015144
3   Tue, Oct 14  0.186603  0.166868 -0.019735  0.125598  0.111245 -0.014353
4   Wed, Oct 15  0.194743  0.168269 -0.026474  0.076464  0.112981  0.036517
5   Thu, Oct 16  0.167679  0.163706 -0.003974  0.099635  0.077411 -0.022224
6   Fri, Oct 17  0.195187  0.162821 -0.032367  0.101604  0.056410 -0.045194
7   Sat, Oct 18  0.174051  0.144172 -0.029879  0.110759  0.095092 -0.015667
8   Sun, Oct 19  0.189580  0.172166 -0.017414  0.086831  0.110473  0.023643
9   Mon, Oct 20  0.191638  0.177907 -0.013731  0.112660  0.113953  0.001294
10  Tue, Oct 21  0.226067  0.165509 -0.060558  0.121107  0.082176 -0.038931
11  Wed, Oct 22  0.193317  0.159800 -0.033517  0.109785  0.087391 -0.022394
12  Thu, Oct 23  0.190977  0.190031 -0.000946  0.084211  0.105919  0.021708
13  Fri, Oct 24  0.326895  0.278336 -0.048559  0.181278  0.134864 -0.046414
14  Sat, Oct 25  0.254703  0.189836 -0.064868  0.185239  0.121076 -0.064163
15  Sun, Oct 26  0.227401  0.220779 -0.006622  0.146893  0.145743 -0.001150
16  Mon, Oct 27  0.306983  0.276265 -0.030718  0.163373  0.154345 -0.009028
17  Tue, Oct 28  0.209239  0.220109  0.010870  0.123641  0.163043  0.039402
18  Wed, Oct 29  0.265223  0.276479  0.011255  0.116373  0.132050  0.015676
19  Thu, Oct 30  0.227520  0.284341  0.056820  0.102180  0.092033 -0.010147
20  Fri, Oct 31  0.246459  0.252078  0.005619  0.143059  0.170360  0.027301
21   Sat, Nov 1  0.229075  0.204317 -0.024758  0.136564  0.143885  0.007321
22   Sun, Nov 2  0.297258  0.251381 -0.045877  0.096681  0.142265  0.045584

Link to p-value for binomial test: https://www.graphpad.com/quickcalcs/binomial1/

In [200]:
print('For Gross conversion, experiment group is higher than the control group for', len(df_sign[df_sign.GC_diff > 0]),
      'out of', len(df_sign))
print('p-value is 0.0026, which is significant.')

For Gross conversion, experiment group is higher than the control group for 4 out of 23
p-value for 0.17391304347826086 is 


In [203]:
print('For Net conversion, experiment group is higher than the control group for', len(df_sign[df_sign.NC_diff > 0]),
      'out of', len(df_sign))
print('p-value is 0.6776, which is not significant.')

For Net conversion, experiment group is higher than the control group for 10 out of 23
p-value is 0.6776, which is not significant.
